### ISSUE TITLE
Inconsistent tablespace paths after running pg_rewind due to alternate timeline

### ISSUE THREAD

**Issue by tvesely** [greenplum-db/gpdb#7772]
### Greenplum version or build
6.0.0_beta.3

I've run into an edge case with pg_rewind that can lead to a primary and mirror segment to have different paths for the same tablespace.

1. Primary and mirror on TID 1.
1. Mirror promoted, TID 2 created.
1. Tablespace with path /ts1 created on TID 2, chooses OID 12000 for the tablespace. This results in this symlink on the old mirror:
  `$OLD_MIRROR_PG_DATA/pg_tblspc/12000 -> /ts1`
[0001-Edge-case-Tablespace-paths-out-of-sync.patch.zip](https://github.com/greenplum-db/gpdb/files/3217702/0001-Edge-case-Tablespace-paths-out-of-sync.patch.zip)

1. Meanwhile, back in TID 1 (the old primary), a tablespace created on with a path /alternate_timeline_tablespace, chooses the OID 12000. This results in the following symlink on the old primary:
  `$OLD_PRIMARY_PG_DATA/pg_tblspc/12000 -> /alternate_timeline_tablespace`
1. pg_rewind runs on old primary. Because it sees that the symlink for the tablespace with oid 12000 already exists, it co-opts this path and populates it with  created in TID 2

What are the implications of having the primary/mirror tablespaces mismatching? Aside from being less than ideal, does this break any major assumptions we have made when designing our mirroring strategy?

If we find that the primary/mirror have different tablespace mappings, should we be throwing an error in pg_rewind? 

### Step to reproduce the behavior

Attached is a patch that triggers the issue

**Comment by asimrp** [comment]
```
# Create a new tablespace in the old master (which should not be replicated on to the old standby)                                                                                                               
PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL -c "CREATE TABLESPACE ts_old_timeline LOCATION '$TS_OLD_TIMELINE_LOCATION';"
```
Using utility mode to connect to old primary, which is already marked down in `gp_segment_configuration` because it's mirror is promoted, does not represent a production use case.  A lot of things can go terribly wrong in Greenplum using utility mode connections to segments.

I'm wondering if this should be considered a legitimate issue that needs to be fixed.

**Comment by BaiShaoqi** [comment]
> 3. Tablespace with path /ts1 created on TID 2, chooses OID 12000 for the tablespace. This results in this symlink on the old mirror:
>    `$OLD_MIRROR_PG_DATA/pg_tblspc/12000 -> /ts1`
> 4. Meanwhile, back in TID 1 (the old primary), a tablespace created on with a path /alternate_timeline_tablespace, chooses the OID 12000. This results in the following symlink on the old primary:

Skip all your questions about `What are the implications of having the primary/mirror tablespaces mismatching? Aside from being less than ideal, does this break any major assumptions we have made when designing our mirroring strategy?`
The problem begins at step 3, it create a tablespace in the old mirror when after promotion but before pg_rewind, but old master did not know about it.  At present, pg_rewind can not handle this situation, quoted from Michael Paquier :
> there maybe a feature for `there is of course the argument of pg_rewind which
> would need an option similar to pg_basebackup's --tablespace-mapping
> so as a target instance does not finish by using the same tablespace
> path as the source where there is a tablespace of difference during
> the operation

**Comment by tvesely** [comment]
> ```
> # Create a new tablespace in the old master (which should not be replicated on to the old standby)                                                                                                               
> PGOPTIONS=${PGOPTIONS_UTILITY} $MASTER_PSQL -c "CREATE TABLESPACE ts_old_timeline LOCATION '$TS_OLD_TIMELINE_LOCATION';"
> ```
> 
> Using utility mode to connect to old primary, which is already marked down in `gp_segment_configuration` because it's mirror is promoted, does not represent a production use case. A lot of things can go terribly wrong in Greenplum using utility mode connections to segments.

That's true, but  is a far more likely scenario is  somebody might be using the master and standby master in the same type of situation. In that case we _could_ run into this issue in production environments.

> 
> I'm wondering if this should be considered a legitimate issue that needs to be fixed.

It's definitely an edge case, but I feel it's still a legitimate issue.

**Comment by tvesely** [comment]
Come to think of it, if it does happen on the master, then it's not going to fill up storage. At worst, it could lead to temporary files going to the wrong place.

**Comment by asimrp** [comment]
@tvesely we do not support incremental recovery for master/standby.  The only option to restore a failed master or standby is `gpinitstandby`, which performs full recovery using `pg_basebackup`.

**Comment by d** [comment]
> Using utility mode to connect to old primary, which is already marked down in `gp_segment_configuration` because it's mirror is promoted, does not represent a production use case. A lot of things can go terribly wrong in Greenplum using utility mode connections to segments.

Asim, I don't think this *only* happens in utility mode. Just because the mirror is promoted doesn't mean that the ex-primary understands that it is demoted (and needs to unwind its work after the timeline divergence -- in fact, this is too much to ask). A tablespace creation could have been in-flight but the network blipped and failover happened.

**Comment by asimrp** [comment]
An in-flight tablespace creation transaction cannot commit on old primary after its mirror is promoted.  However, I realized that we are not concerned with commit status of such a transaction but whether the transaction managed to create a symlink in data directory.  And that directory can be created on old primary after it has been marked down, as you pointed out.

I stand corrected, this is a legitimate issue.

Root cause of this issue is the Greenplum-specific deviation from PostgreSQL to include dbid in tablespace path.  However, the only use case that deviation seems to achieve is ability to create tablespaces in a demo cluster.  No production deployment will ever have a primary and mirror deployed on the same hardware.  So, I'm tempted to ask whether it is worthwhile to incur such a diff against upstream, and that has such adverse implications on production deployments?  Shouldn't we invent alternatives in developer workflow and CI for `create tablespace` to work on a demo cluster?
